### PR TITLE
[Chore] #1397 원장 해시 producer + admin review record 생성기

### PR DESCRIPTION
**Pull request by AquilaXk**
## 관련 이슈

Refs #1397

이 PR은 #1397 DoD 중 "build-spec에 sourceSnapshotSetHash, alias ledger hash, facility/route evidence ledger hash, override hash가 연결된다"와 "admin review 승인 record 생성"의 선행 도구를 채운다. #1397은 1차 승격 실행(live sample 수집·inventory 승격)이 남아 있어 닫지 않는다.

## 작업 배경

- 수도권 신 도식 데이터팩의 첫 production 게시(#1950 작업 8)가 admission summary의 해시 6종에 걸려 있다.
- 정본 계약 `tools/datapack/source-admission-runbook.json`의 `requiredAdminReviewFields`는 licenseEvidenceHash·aliasLedgerHash·operatorMappingLedgerHash·facilityEvidenceLedgerHash·routeEvidenceLedgerHash·overrideHash를 요구하지만, 지금까지 이 해시들의 **consumer만 존재**했다(`run-source-admission-pipeline.mjs`의 형식 검증). producer가 없어 admin review record의 원장 해시를 손으로 채워야 했고, 이는 위조 가능한 경로였다.

## 작업 내용

- **`tools/datapack/export-ledger-hashes.mjs`** — 원장 해시 exporter. 6종(alias·operator-mapping·facility-evidence·route-evidence·override·license)의 canonical 직렬화 + sha256을 산출한다.
  - 결정성 규칙: 객체 key 재귀 사전순 정렬 → 공백 없는 `JSON.stringify` → sha256 hex. 레코드 집합은 canonical row 문자열로 직렬화 후 사전순 정렬 → 입력 순서와 무관하게 동일 해시.
  - canonical 소스(실측): fixture pack의 `stationAliases`(alias)/`operators`(operator-mapping)/`stationFacilityEvidence`(있으면) 또는 `facilities`(facility-evidence)/`networkEdges`(route-evidence), manual override ledger 파일(override), `source-inventory.json`의 source `license` 블록(license).
- **`tools/datapack/build-admin-review-record.mjs`** — admin review record 생성기. runbook `requiredAdminReviewFields`를 채운다.
  - 위조 방지: hash 6종은 exporter의 JSON 산출물 파일에서만 읽고 `kind`를 검증한다(손으로 hex를 인자로 끼워넣을 수 없다). `sampleEvidenceHash`는 sample evidence의 `evidenceHash`에서, 승인자·결정은 입력 인자.
  - `run-source-admission-pipeline.mjs`의 `validateAdminReview` 계약(필드 집합·quotaEvidence 키·productionSource 규칙)과 일치.
- 신규 도구는 기존 consumer(`run-source-admission-pipeline.mjs`)를 변경하지 않고 그 입력을 생성만 한다.

## 검증

- 실행한 명령과 결과:
  - `node --test tools/datapack/datapack-tools.test.mjs` → 226 pass, 0 fail (신규 9 테스트 포함)
  - `node --test tools/ci/repository-contract.test.mjs` → 179 pass, 0 fail
  - `git diff --cached --check` → clean
  - end-to-end: exporter → 생성기 → 실 `run-source-admission-pipeline.mjs` 통과, summary가 exporter 원장 해시를 그대로 옮기는 것을 확인

## 검증 증거

| 항목 | 플랫폼 | 확인 방법 | 증거 | 결과 |
| --- | --- | --- | --- | --- |
| exporter 6종 산출·결정성·순서 불변 | tools/datapack | `node --test --test-name-pattern "원장 해시"` | datapack-tools.test.mjs 신규 테스트 | PASS |
| 생성기 필수 필드·위조 방지·decision | tools/datapack | `node --test --test-name-pattern "admin review record"` | datapack-tools.test.mjs 신규 테스트 | PASS |
| exporter→생성기→실 pipeline e2e | tools/datapack | `node --test --test-name-pattern "그대로 통과"` | datapack-tools.test.mjs 신규 테스트 | PASS |
| 미승인 admission negative gate 유지 | tools/datapack | `node --test --test-name-pattern "admin 승인 없는"` | 기존 테스트 green | PASS |
| 기존 계약 회귀 없음 | tools/ci | `node --test tools/ci/repository-contract.test.mjs` | 179 pass | PASS |

## Version impact

- [ ] no version change
- [ ] mobile patch
- [ ] mobile minor
- [ ] mobile major
- [ ] backend deploy only
- [x] datapack release only
- [ ] route/realtime contract change
- [ ] DB migration change

datapack admission 도구만 추가하며 데이터팩 아티팩트나 버전 자체는 이 PR에서 변경하지 않는다. 실제 datapack version bump는 후속 1차 승격·게시(#1950 작업 8)에서 발생한다.

## Route commercialization gate impact

- [x] route-commercialization-gate.json 영향 없음
- [ ] route ETA accuracy, realtime coverage, accessibility regression, route v2 contract report를 갱신했다.
- [ ] 상용 경로/ETA claim을 추가하거나 변경하지 않는다.

## Route release readiness tracker impact

- [x] route-release-readiness-tracker.json 영향 없음
- [ ] #1414 하위 release blocker issue 또는 production evidence 완료 조건을 갱신했다.
- [ ] 실시간/교통약자 길찾기 출시 준비 완료 claim을 추가하거나 변경하지 않는다.

### Version decision

- mobile versionName: 변경 없음
- mobile versionCode: 변경 없음
- datapack version: 변경 없음 (도구 추가만, 아티팩트 미생성)
- route contract: 변경 없음
- realtime contract: 변경 없음
- backend identity: 변경 없음

## 리뷰어 메모

- 리뷰어가 먼저 봐야 할 지점:
  - `export-ledger-hashes.mjs`의 canonical 직렬화·정렬 규칙(`canonicalizeRows`, `sortJson`)이 결정적이고 순서 불변인지.
  - `build-admin-review-record.mjs`의 위조 방지 경로: hash가 exporter 산출물 파일에서만 읽히고 `kind`가 검증되는지(`readLedgerHash`).
  - 생성기 출력이 `run-source-admission-pipeline.mjs`의 `validateAdminReview` 계약과 필드·quotaEvidence 키까지 일치하는지 — e2e 테스트로 고정했다.

## 리스크

- facility-evidence 원장은 fixture에 `stationFacilityEvidence`가 없으면 `facilities`로 폴백한다. capital 팩이 evidence row를 실제로 갖게 되면 exporter가 자동으로 evidence row 기준으로 전환되므로 해시가 바뀐다(의도된 provenance 반영). 후속 게시에서 실 데이터로 재산출 필요.
- 승인자·결정은 인자로 받으므로 운영자가 실제 승인 결정을 정확히 입력할 책임이 남는다(도구는 형식·위조만 막는다).

## 체크리스트

- [x] PR 본문은 이 템플릿 섹션을 삭제하지 않고 모두 채웠다.
- [ ] CI 결과를 확인했다.
- [ ] CodeRabbit 리뷰를 확인했다.
- [ ] GitHub PR Review 객체가 있는지 확인했다. CodeRabbit status check만으로는 리뷰 완료로 보지 않는다.
- [ ] CodeRabbit 실행이 불가능하거나 PR Review 객체가 없으면 Codex CLI code review를 단일 PR review로 게시했다.
- [ ] 배포 영향이 있는 경우 CD 상태를 확인했다.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tools/datapack/export-ledger-hashes.mjs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      }
+      return requiredArray(pack.facilities ?? [], "pack.facilities").map((row) => ({
+        id: requiredString(row.id, "facilities.id"),
+        stationId: requiredString(row.stationId, "facilities.stationId"),
+        type: requiredString(row.type, "facilities.type"),
+        status: requiredString(row.status, "facilities.status"),
+      }));
+    }
+    case "route-evidence":
+      return requiredArray(pack.networkEdges ?? [], "pack.networkEdges").map((row) => ({
```
Comment (AquilaXk) [inline tools/datapack/export-ledger-hashes.mjs]: _🎯 Functional Correctness_ | _🟠 Major_ | _⚡ Quick win_

**facility-evidence 폴백이 산출물에 표기되지 않아 잘못된 원장을 조용히 해시할 수 있습니다.**

`ledgerRowsForPack`의 facility-evidence 분기는 `pack.stationFacilityEvidence`가 비면 스키마가 전혀 다른 `pack.facilities`로 조용히 폴백하는데, 산출 객체에 어느 소스를 해시했는지 기록되지 않습니다(실측: 소스 구분 필드 없음). 폴백 판정이 pack 단위 루프 안이라 pack 간 서로 다른 스키마의 row가 한 해시로 병합될 수도 있습니다. 이 값은 production admission attestation에 봉인되므로, 리뷰어·파이프라인이 "진짜 evidence 해시"와 "폴백 해시"를 구별할 수 없는 것은 attestation 무결성 문제입니다. 산출물에 `evidenceSource` 표기를 추가하고 pack 간 혼합 폴백은 거부하세요.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@tools/datapack/export-ledger-hashes.mjs` around lines 96 - 115, the facility-evidence kind silently falls back from pack.stationFacilityEvidence to pack.facilities (a different schema) without recording which source produced the hash. Add an evidenceSource marker to the returned object so downstream review/attestation can detect a fallback, and consider rejecting mixed per-pack fallback.
```

</details>

_Source: Claude Code review_

File: tools/datapack/export-ledger-hashes.mjs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  if (overrides.ledgerSource !== "manual_overrides") {
+    throw new Error("override ledger ledgerSource must be manual_overrides");
+  }
+  return {
+    schemaVersion: 1,
+    artifactKind: "datapack-override-ledger-hash",
+    kind: "override",
+    overridesPath: path.relative(root, overridesPath),
+    rowCount: Array.isArray(overrides.facilityStatusUpdates) ? overrides.facilityStatusUpdates.length : 0,
+    ledgerHash: sha256(JSON.stringify(sortJson(overrides))),
```
Comment (AquilaXk) [inline tools/datapack/export-ledger-hashes.mjs:216]: _🎯 Functional Correctness_ | _🟡 Minor_ | _⚡ Quick win_

**override/license 해시는 canonicalizeRows를 거치지 않아 배열 순서에 민감합니다 — doc-comment의 "입력 순서 무관" 주장과 어긋납니다.**

override는 원장 객체 전체를 `sortJson`만 통과시키는데 `sortJson`은 key만 정렬하고 배열 원소 순서는 보존합니다(실측). 같은 `facilityStatusUpdates` 항목을 다른 순서로 기록하면 overrideHash가 달라집니다. 현재 fixture는 빈 배열이라 테스트가 이를 노출하지 못합니다. 배열도 canonical 정렬을 적용하거나 doc-comment의 결정성 범위를 row 기반 kind로 한정 정정하세요.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@tools/datapack/export-ledger-hashes.mjs` around line 165, the override hash uses sortJson only, which does not sort array elements (e.g. facilityStatusUpdates order is preserved), contradicting the determinism doc-comment. Either canonicalize array order or correct the comment to scope order-independence to the row-based kinds only.
```

</details>

_Source: Claude Code review_

File: tools/datapack/lib/ledger-admission-cli.mjs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  return value;
+}
+
+// 사전순 비교자 — 문자열 key/값 정렬에 재사용.
+export function compareStrings(left, right) {
+  if (left < right) return -1;
+  if (left > right) return 1;
+  return 0;
+}
```
Comment (AquilaXk) [inline tools/datapack/lib/ledger-admission-cli.mjs:47]: _📐 Maintainability & Code Quality_ | _🟡 Minor_ | _💤 Low value_

**공유 헬퍼의 `compareStrings`(code-point)와 admission 파이프라인의 `sortJson`(localeCompare) canonical 순서가 불일치합니다.**

두 구현이 같은 이름(`sortJson`)으로 canonical 직렬화를 의도하지만 비교자가 달라 대소문자·숫자·악센트 등에서 순서가 광범위하게 갈립니다(실측). 현재는 각 해시의 producer가 단일해 실피해가 없으나(파이프라인은 ledger 해시를 recompute 없이 복사), 향후 consumer가 파이프라인 `sortJson`으로 재검증하면 조용히 mismatch가 납니다. 단일 canonical 비교자로 통일을 권고합니다.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@tools/datapack/lib/ledger-admission-cli.mjs` around lines 43 - 47, compareStrings uses code-point ordering while run-source-admission-pipeline.mjs sortJson uses localeCompare, so the two canonical serializations diverge for many key sets. Unify them on a single canonical comparator to prevent a future consumer from recomputing a mismatching hash.
```

</details>

_Source: Claude Code review_

**Review by AquilaXk — COMMENTED** [review]
**Actionable comments posted: 4**

> [!CAUTION]
> Some comments are outside the diff and can’t be posted inline due to platform limitations.
>
> <details>
> <summary>⚠️ Outside diff range comments (1)</summary><blockquote>
>
> <details>
> <summary>tools/datapack/datapack-tools.test.mjs (1)</summary><blockquote>
>
> `facility-evidence 테스트 블록`: _🎯 Functional Correctness_ | _🟡 Minor_ | _⚡ Quick win_
>
> **facility-evidence primary 분기(stationFacilityEvidence) 테스트가 전무합니다.**
>
> 모든 facility 테스트가 `catalog-fixture.json`(stationFacilityEvidence 0행 실측)을 사용해 의도된 production 원장 경로는 한 번도 실행되지 않고 폴백만 검증됩니다. stationFacilityEvidence가 채워진 fixture로 primary 분기·결정성·폴백과의 해시 구분을 검증하는 케이스를 추가하세요.
>
> <details>
> <summary>🤖 Prompt for AI Agents</summary>
>
> ```
> Verify each finding against current code. Fix only still-valid issues, skip the
> rest with a brief reason, keep changes minimal, and validate.
>
> In `@tools/datapack/datapack-tools.test.mjs`, add a facility-evidence test that uses a fixture with non-empty pack.stationFacilityEvidence so the primary (non-fallback) branch is exercised, asserting deterministic order-independent hashing and that it differs from the facilities fallback.
> ```
>
> </details>
>
> _Source: Claude Code review_
>
> </blockquote></details>
>
> </blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

Inline comments:
In `@tools/datapack/export-ledger-hashes.mjs` around lines 96 - 115, the facility-evidence kind silently falls back from pack.stationFacilityEvidence to pack.facilities (a different schema) without recording which source produced the hash. Add an evidenceSource marker to the returned object so downstream review/attestation can detect a fallback, and consider rejecting mixed per-pack fallback.

In `@tools/datapack/export-ledger-hashes.mjs` around line 165, the override hash uses sortJson only, which does not sort array elements (e.g. facilityStatusUpdates order is preserved), contradicting the determinism doc-comment. Either canonicalize array order or correct the comment to scope order-independence to the row-based kinds only.

In `@tools/datapack/lib/ledger-admission-cli.mjs` around lines 43 - 47, compareStrings uses code-point ordering while run-source-admission-pipeline.mjs sortJson uses localeCompare, so the two canonical serializations diverge for many key sets. Unify them on a single canonical comparator to prevent a future consumer from recomputing a mismatching hash.

---

Outside diff comments:
In `@tools/datapack/datapack-tools.test.mjs`, add a facility-evidence test that uses a fixture with non-empty pack.stationFacilityEvidence so the primary (non-fallback) branch is exercised, asserting deterministic order-independent hashing and that it differs from the facilities fallback.
```

</details>

<details>
<summary>🪄 Autofix (Beta)</summary>

Claude Code fallback에는 CodeRabbit Autofix가 없습니다. 각 finding을 현재 코드에서 검증한 뒤 별도 승인된 수정 흐름으로 처리합니다.

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: Path: .coderabbit.yaml

**Review profile**: ASSERTIVE

**Plan**: Claude Code fallback

**Run ID**: `N/A (Claude Code fallback)`

**Review source**: Claude Code CLI

**Repository visibility**: public GitHub repository

**Trigger**: CodeRabbit rate limit — Review limit reached(Review 객체 미생성)로 인한 폴백 리뷰

**Base**: `c8259681104a48c6fec615205a65d28ab32fdfb2`

**Head**: `8c7f2066efea7473d9e039f4373a1f9983fa1b79`

**Command**: `git diff origin/main...origin/feat/1397-ledger-admission` (격리 detached worktree에서 `node --test`·결정성/위조 방지 실측 병행)

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between c8259681104a48c6fec615205a65d28ab32fdfb2 and 8c7f2066efea7473d9e039f4373a1f9983fa1b79.

</details>

<details>
<summary>📒 Files selected for processing (4)</summary>

* `tools/datapack/export-ledger-hashes.mjs`
* `tools/datapack/build-admin-review-record.mjs`
* `tools/datapack/lib/ledger-admission-cli.mjs`
* `tools/datapack/datapack-tools.test.mjs`

</details>

</details>

<details>
<summary>📜 Review details</summary>

<details>
<summary>🔇 Additional comments (5)</summary><blockquote>

<details>
<summary>tools/datapack/build-admin-review-record.mjs (3)</summary><blockquote>

`55-71`: LGTM!

`73-81`: LGTM!

`100-131`: LGTM!

</blockquote></details>

<details>
<summary>tools/datapack/export-ledger-hashes.mjs (1)</summary><blockquote>

`170-171`: LGTM!

</blockquote></details>

<details>
<summary>tools/datapack/lib/ledger-admission-cli.mjs (1)</summary><blockquote>

`13-22`: LGTM!

</blockquote></details>

</blockquote></details>

</details>

File: tools/datapack/export-ledger-hashes.mjs (hunk 1 above)
Comment (AquilaXk) [reply tools/datapack/export-ledger-hashes.mjs]: 8356b947에서 수정 — `exportFacilityEvidenceLedgerHash` 분기를 신설해 산출물에 `evidenceSource: "stationFacilityEvidence" | "facilities"`를 표기하고, pack 간 source 혼합은 에러로 거부합니다. consumer(`readLedgerHash`)는 kind·ledgerHash만 읽어 계약 호환(smoke run으로 해시 불변 확인). primary 분기 테스트 3종도 추가(임시 fixture 관례) — 결정성·폴백 해시와의 구분·evidenceSource 표기 검증. 검증: `node --test tools/datapack/datapack-tools.test.mjs` 229/0.

File: tools/datapack/export-ledger-hashes.mjs (hunk 2 above)
Comment (AquilaXk) [reply tools/datapack/export-ledger-hashes.mjs:216]: 8356b947에서 doc-comment 정정으로 처리 — 실측 결과 `apply-admin-review-overrides.mjs`의 `latestFacilityUpdates`가 reviewedAt 동률 시 배열 인덱스를 tiebreaker로 사용해 `facilityStatusUpdates` 순서가 의미 있는 데이터입니다. 배열 canonical 정렬을 적용하면 override 의미(최신 판정)가 깨지므로, doc-comment에 "객체 key만 정렬·배열 순서는 원본 보존이 계약"임을 명시하고 순서 민감성 테스트(reverse → 해시 변동)를 추가했습니다.

File: tools/datapack/lib/ledger-admission-cli.mjs (hunk 3 above)
Comment (AquilaXk) [reply tools/datapack/lib/ledger-admission-cli.mjs:47]: 8356b947에서 수정 — pipeline(`run-source-admission-pipeline.mjs`)의 localeCompare 3곳을 공유 lib의 code-point 비교자(`compareStrings`/`sortJson`)로 통일했습니다(ICU 비의존 결정성, #1965 선례). 호환성 실측: 현재 데이터(adminReview 키셋·inventory source id 전체·quotaEvidence 키)에서 두 정렬 결과가 100% 동일해 기록된 `adminReviewRecordHash`·`sourceInventorySha256` 불변. e2e(exporter→record→pipeline) green, localeCompare 잔존 0.
